File: lib/clients/openalex.ts
import type { Paper, Author, PaperSearchResult } from "../types/paper";

const BASE_URL = "https://api.openalex.org";
const POLITE_EMAIL = process.env.OPENALEX_EMAIL || "";

function buildUrl(path: string, params: Record<string, string>): string {
  const url = new URL(`${BASE_URL}${path}`);
  for (const [key, value] of Object.entries(params)) {
    url.searchParams.set(key, value);
  }
  if (POLITE_EMAIL) {
    url.searchParams.set("mailto", POLITE_EMAIL);
  }
  return url.toString();
}

// OpenAlex returns abstract as an "inverted index" — reconstruct it
function reconstructAbstract(
  invertedIndex: Record<string, number[]> | null | undefined
): string | null {
  if (!invertedIndex) return null;
  const words: [number, string][] = [];
  for (const [word, positions] of Object.entries(invertedIndex)) {
    for (const pos of positions) {
      words.push([pos, word]);
    }
  }
  words.sort((a, b) => a[0] - b[0]);
  return words.map(([, word]) => word).join(" ");
}

/* eslint-disable @typescript-eslint/no-explicit-any */
function normalizePaper(work: any): Paper {
  return {
    id: work.id?.replace("https://openalex.org/", "") ?? "",
    title: work.title ?? "Untitled",
    abstract: reconstructAbstract(work.abstract_inverted_index),
    authors: (work.authorships ?? []).map(
      (a: any): Author => ({
        name: a.author?.display_name ?? "Unknown",
        institution:
          a.institutions?.[0]?.display_name ?? null,
      })
    ),
    year: work.publication_year ?? null,
    citationCount: work.cited_by_count ?? 0,
    fields: (work.topics ?? [])
      .slice(0, 3)
      .map((t: any) => t.display_name),
    doi: work.doi ?? null,
    pdfUrl:
      work.open_access?.oa_url ??
      work.primary_location?.pdf_url ??
      null,
    openAccess: work.open_access?.is_oa ?? false,
    openalexId: work.id ?? "",
  };
}
/* eslint-enable @typescript-eslint/no-explicit-any */

export interface SearchOptions {
  query: string;
  page?: number;
  perPage?: number;
  yearFrom?: number;
  yearTo?: number;
  openAccessOnly?: boolean;
  sort?: string; // e.g. "relevance_score:desc", "cited_by_count:desc", "publication_year:desc"
}

export async function searchPapers(opts: SearchOptions): Promise<PaperSearchResult> {
  const { query, page = 1, perPage = 20, yearFrom, yearTo, openAccessOnly, sort = "relevance_score:desc" } = opts;

  // Build filter string
  const filters: string[] = [];
  if (yearFrom && yearTo) {
    filters.push(`publication_year:${yearFrom}-${yearTo}`);
  } else if (yearFrom) {
    filters.push(`publication_year:${yearFrom}-`);
  } else if (yearTo) {
    filters.push(`publication_year:-${yearTo}`);
  }
  if (openAccessOnly) {
    filters.push("open_access.is_oa:true");
  }

  const params: Record<string, string> = {
    search: query,
    page: String(page),
    per_page: String(perPage),
    sort,
  };
  if (filters.length > 0) {
    params.filter = filters.join(",");
  }

  const url = buildUrl("/works", params);

  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`OpenAlex search failed: ${res.status}`);
  }

  const data = await res.json();
  return {
    papers: (data.results ?? []).map(normalizePaper),
    totalCount: data.meta?.count ?? 0,
    page,
    perPage,
  };
}

export async function getPaper(openalexId: string): Promise<Paper | null> {
  // Accept both full URL and short ID
  const id = openalexId.startsWith("https://")
    ? openalexId
    : `https://openalex.org/${openalexId}`;

  const url = buildUrl(`/works/${encodeURIComponent(id)}`, {});

  const res = await fetch(url);
  if (!res.ok) {
    if (res.status === 404) return null;
    throw new Error(`OpenAlex get paper failed: ${res.status}`);
  }

  const data = await res.json();
  return normalizePaper(data);
}

File: pipeline/clients/openalex.py
from __future__ import annotations

import os
from typing import Any, Optional

import requests

from models.paper import Author, Paper

_BASE_URL = "https://api.openalex.org"


# ── 내부 헬퍼 ──────────────────────────────────────────────────────────────


def _base_params() -> dict[str, str]:
    """mailto polite-pool 파라미터 (환경변수 있으면 포함)."""
    email = os.getenv("OPENALEX_EMAIL", "")
    return {"mailto": email} if email else {}


def _reconstruct_abstract(inverted: dict[str, list[int]] | None) -> Optional[str]:
    """OpenAlex abstract_inverted_index → 평문 변환."""
    if not inverted:
        return None
    words: list[tuple[int, str]] = [
        (pos, word)
        for word, positions in inverted.items()
        for pos in positions
    ]
    words.sort()
    return " ".join(w for _, w in words)


def _normalize(work: dict[str, Any]) -> Paper:
    authorships = work.get("authorships") or []
    topics = work.get("topics") or []
    oa = work.get("open_access") or {}
    primary = work.get("primary_location") or {}

    return Paper(
        id=work.get("id", "").replace("https://openalex.org/", ""),
        title=work.get("title") or "Untitled",
        abstract=_reconstruct_abstract(work.get("abstract_inverted_index")),
        authors=[
            Author(
                name=(a.get("author") or {}).get("display_name", "Unknown"),
                institution=(
                    (a.get("institutions") or [{}])[0].get("display_name")
                    if a.get("institutions")
                    else None
                ),
            )
            for a in authorships
        ],
        year=work.get("publication_year"),
        citation_count=work.get("cited_by_count", 0),
        fields=[t["display_name"] for t in topics[:3] if t.get("display_name")],
        doi=work.get("doi"),
        pdf_url=oa.get("oa_url") or primary.get("pdf_url"),
        open_access=oa.get("is_oa", False),
    )


# ── 공개 API ───────────────────────────────────────────────────────────────


def fetch_paper(openalex_id: str) -> Optional[Paper]:
    """OpenAlex ID 또는 전체 URL로 논문 단건 조회.

    Args:
        openalex_id: "W2741809807" 또는 "https://openalex.org/W2741809807"
    Returns:
        Paper 객체, 없으면 None.
    """
    if not openalex_id.startswith("https://"):
        openalex_id = f"https://openalex.org/{openalex_id}"

    res = requests.get(
        f"{_BASE_URL}/works/{openalex_id}",
        params=_base_params(),
        timeout=15,
    )
    if res.status_code == 404:
        return None
    res.raise_for_status()
    return _normalize(res.json())


def search_papers(
    query: str,
    per_page: int = 10,
    year_from: Optional[int] = None,
    year_to: Optional[int] = None,
    open_access_only: bool = False,
    sort: str = "relevance_score:desc",
) -> list[Paper]:
    """키워드로 OpenAlex 논문 검색.

    Args:
        query: 검색어 (영문 권장).
        per_page: 반환 논문 수 (최대 200).
        year_from / year_to: 연도 범위 필터.
        open_access_only: True면 OA 논문만.
        sort: 정렬 기준.
    Returns:
        Paper 리스트.
    """
    filters: list[str] = []
    if year_from and year_to:
        filters.append(f"publication_year:{year_from}-{year_to}")
    elif year_from:
        filters.append(f"publication_year:{year_from}-")
    elif year_to:
        filters.append(f"publication_year:-{year_to}")
    if open_access_only:
        filters.append("open_access.is_oa:true")

    params: dict[str, Any] = {
        **_base_params(),
        "search": query,
        "per_page": per_page,
        "sort": sort,
    }
    if filters:
        params["filter"] = ",".join(filters)

    res = requests.get(f"{_BASE_URL}/works", params=params, timeout=15)
    res.raise_for_status()
    return [_normalize(w) for w in res.json().get("results", [])]
